ads: fix adsc version nonce handling

ramaraochavali opened the pull request:
Fixes adsc version and nonce handling and removes the workarounds in the ads code

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | 459f6ccf43085e51ecb3658bf6772bf505ba7bce | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/25418/unit-tests_istio/16297) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented on pkg/adsc/adsc.go:
```diff
@@ -551,12 +522,10 @@ func (a *ADSC) handleRecv() {
 		}
 		a.Received[msg.TypeUrl] = msg
-		// TODO: add hook to inject nacks
-		a.ack(msg)
```
where is the ACK now? It seems missing entirely?

ramaraochavali replied:
Ack is not needed when you respond with next request as is being done below.

howardjohn replied:
According to what?

> The xDS client should ACK or NACK every DiscoveryResponse received from the management server.

And I confirmed that Envoy behaves this way

Even if our tests pass we probably want to simulate envoy as closely as possible?

ramaraochavali replied:
https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#aggregated-discovery-service - For ADS Ack is the next EDS request for CDS response.

ramaraochavali replied:
I can change this to send ACK for EDS/RDS to match exactly what Envoy expects in ADS. But for CDS/LDS - Envoy follows it with an EDS request/RDS request

howardjohn commented:
Interesting... Thanks for the link. We still need CDS ack for pushes though
right? it's just the first request sequence where it's skipped

On Fri, Jul 10, 2020, 9:17 PM Rama Chavali <notifications@github.com> wrote:

> *@ramaraochavali* commented on this pull request.
> ------------------------------
>
> In pkg/adsc/adsc.go
> <https://github.com/istio/istio/pull/25418#discussion_r453152921>:
>
> >  			a.sync[gt.String()] = time.Now()
>  		}
>  		a.Received[msg.TypeUrl] = msg
> -		// TODO: add hook to inject nacks
> -		a.ack(msg)
>
> I can change this to send ACK for EDS/RDS to match exactly what Envoy
> expects in ADS. But for CDS/LDS - Envoy follows it with an EDS request/RDS
> request
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/25418#discussion_r453152921>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEYGXINPZYQV4NE5OK7CADR27RVDANCNFSM4OWSDRIQ>
> .
>

howardjohn commented:
```
2020-07-11T04:28:46.653412Z     info    ads     ADS:CDS: REQ sidecar~10.0.0.14~.~.svc.cluster.local-1 version:
2020-07-11T04:28:46.659095Z     info    ads     CDS: PUSH for node:. clusters:18 services:4 version:2020-07-10T21:28:41-07:00/2
2020-07-11T04:28:46.674997Z     debug   ads     ADS:EDS: REQ sidecar~10.0.0.14~.~.svc.cluster.local-1 clusters:14
2020-07-11T04:28:46.675571Z     info    ads     EDS: PUSH for node:. clusters:14 endpoints:0 empty:14
2020-07-11T04:28:46.675597Z     debug   ads     ADS:CDS: ACK sidecar~10.0.0.14~.~.svc.cluster.local-1 2020-07-10T21:28:41-07:00/2 jUoYSiGPijQ=b2c9ac52-ae88-4476-bd9e-9567ed863c8c
2020-07-11T04:28:46.677509Z     debug   ads     ADS:EDS: ACK sidecar~10.0.0.14~.~.svc.cluster.local-1 2020-07-10T21:28:41-07:00/2 jUoYSiGPijQ=b43a90e5-e6ab-4017-a357-473e34490f0b
2020-07-11T04:28:46.926246Z     debug   ads     ADS:LDS: REQ sidecar~10.0.0.14~.~.svc.cluster.local-1
2020-07-11T04:28:46.927792Z     info    ads     LDS: PUSH for node:. listeners:16
2020-07-11T04:28:46.993721Z     debug   ads     ADS:RDS: REQ sidecar~10.0.0.14~.~.svc.cluster.local-1 routes:6
2020-07-11T04:28:46.994181Z     info    ads     RDS: PUSH for node:. routes:6
2020-07-11T04:28:46.994199Z     debug   ads     ADS:LDS: ACK sidecar~10.0.0.14~.~.svc.cluster.local-1 2020-07-10T21:28:41-07:00/2 jUoYSiGPijQ=5f00cddd-20f9-4657-91dc-84efb3673280
2020-07-11T04:28:46.997428Z     debug   ads     ADS:RDS: ACK sidecar~10.0.0.14~.~.svc.cluster.local-1 2020-07-10T21:28:41-07:00/2 jUoYSiGPijQ=dd65f07d-2948-4610-82c9-e3c45c34fc32
```

I see acks for everything?

ramaraochavali commented:
That is interesting. Let me check

ramaraochavali commented:
Confirmed Envoy sends ACK for every request even ADS (including initial setup). Sorry for the confusion. Added the ack back and fixed ack code to send resource names for EDS. We need to fix it for Routes later. Added a TODO.

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | ba3209eca161b3c404d0f23c5062ee05c65f016b | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/25418/unit-tests_istio/16494) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->
